src/core/middlewares/regquota/handler.go

```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
-					return
-				}
-				req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+		if mediaType == schema1.MediaTypeManifest ||
+			mediaType == schema1.MediaTypeSignedManifest ||
+			mediaType == schema2.MediaTypeManifest {
+
+			tagLock, err := rqh.tryLockTag()
+			if err != nil {
+				log.Warningf("Error occurred when to lock tag %s:%s with digest %v", repository, tag, err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to lock tag %s:%s with digest %v", repository, tag, err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.TagLock = tagLock
+
+			data, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to copy manifest body %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to decode manifest body %v", err)), http.StatusInternalServerError)
+				return
+			}
+			req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
 
-				_, desc, err := distribution.UnmarshalManifest(mediaType, data)
-				if err != nil {
-					log.Warningf("Error occurred when to Unmarshal Manifest %v", err)
-					http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
-					return
+			manifest, desc, err := distribution.UnmarshalManifest(mediaType, data)
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to Unmarshal Manifest %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Refrerence = manifest.References()
+			rqh.mfInfo.Digest = desc.Digest.String()
+			rqh.mfInfo.Size = desc.Size
+
+			projectID, err := rqh.getProjectID(strings.Split(repository, "/")[0])
+			if err != nil {
+				log.Warningf("Error occurred when to get project ID %v", err)
+				return
+			}
+			rqh.mfInfo.ProjectID = projectID
+
+			exist, af, err := rqh.imageExist()
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to check Manifest existence %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to check Manifest existence %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Exist = exist
+			if exist {
+				if af.Digest != rqh.mfInfo.Digest {
+					rqh.mfInfo.DigestChanged = true
 				}
-				mfDigest = desc.Digest.String()
-				mfSize = desc.Size
-				log.Infof("manifest digest... %s", mfDigest)
-				log.Infof("manifest size... %v", mfSize)
 			}
+			quotaRes, err := rqh.tryRequireQuota()
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Errorf("Cannot get quota for the manifest %v", err)
+				http.Error(rw, util.MarshalError("StatusNotAcceptable", fmt.Sprintf("Cannot get quota for the manifest %v", err)), http.StatusNotAcceptable)
```

Comment:
How can we determine whether the error is an `internal server error `or `quota not meet the requirement error`?

Reply:
define an new error